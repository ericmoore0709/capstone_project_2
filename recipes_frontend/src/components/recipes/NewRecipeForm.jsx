import { Button, Form, FormGroup, FormText, Input, Label, ListGroup, ListGroupItem } from 'reactstrap';
import PropTypes from 'prop-types';
import { useContext, useState } from 'react';
import DOMPurify from 'dompurify';
import { AuthContext } from '../../contexts/AuthContext';

const NewRecipeForm = ({ addRecipe, errors = [] }) => {

    const INITIAL_FORM_DATA = {
        title: '',
        description: '',
        image: '',
        visibility_id: 1, // default to 'Public'
    };

    const { signedInUser } = useContext(AuthContext);

    const [formData, setFormData] = useState(INITIAL_FORM_DATA);

    const handleInputChange = (e) => {
        const { name, value } = e.target;
        setFormData({ ...formData, [name]: value });
    };

    const handleFormSubmit = (e) => {
        e.preventDefault();

        // Sanitize input fields before submission
        const sanitizedData = {
            title: DOMPurify.sanitize(formData.title),
            description: DOMPurify.sanitize(formData.description),
            image: DOMPurify.sanitize(formData.image),
            visibility_id: +formData.visibility_id,
            author_id: signedInUser.id
        };

        addRecipe(sanitizedData);
        setFormData(INITIAL_FORM_DATA);
    };

    return (
        <div>
            <h2 className='text-center mt-2'>Create a Recipe</h2>
            <Form onSubmit={handleFormSubmit} className='w-50 mx-auto p-2 border'>
                <FormGroup>
                    <Label htmlFor='title'>Title</Label>
                    <Input id='title' name='title' value={formData.title} onChange={handleInputChange} />
                    <FormText>Give your recipe a name!</FormText>
                </FormGroup>
                <FormGroup>
                    <Label htmlFor='description'>Description</Label>
                    <Input id='description' name='description' value={formData.description} onChange={handleInputChange} />
                    <FormText>Give a light description of your delicious masterpiece!</FormText>
                </FormGroup>
                <FormGroup>
                    <Label htmlFor='image'>Image URL</Label>
                    <Input id='image' name='image' value={formData.image} onChange={handleInputChange} />
                    <FormText>Show us an image of your finished meal.</FormText>
                </FormGroup>
                <FormGroup>
                    <Label htmlFor='visibility_id'>Visibility</Label>
                    <Input type='select' name='visibility_id' id='visibility_id' value={formData.visibility_id} onChange={handleInputChange}>
                        <option value="1">Public</option>
                        <option value="2">Community</option>
                        <option value="3">Private</option>
                    </Input>
                    <FormText>Select who can view your recipe.</FormText>
                </FormGroup>
                <FormGroup className='text-center'>
                    <Button color='primary' className='w-25'>Add</Button>
                </FormGroup>
            </Form>

            {(errors.length > 0) && (
                <div>
                    <ListGroup color='danger' className='w-50 mx-auto'>
                        {errors.map((err, index) => (
                            <ListGroupItem key={index} color='danger'>{err}</ListGroupItem>
                        ))}
                    </ListGroup>
                </div>
            )}
        </div>
    );
};

NewRecipeForm.propTypes = {
    addRecipe: PropTypes.func.isRequired,
    errors: PropTypes.array,
    signedInUser: PropTypes.object
};

export default NewRecipeForm;
