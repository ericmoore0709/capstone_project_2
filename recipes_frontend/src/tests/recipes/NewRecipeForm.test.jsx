import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import NewRecipeForm from "../../components/recipes/NewRecipeForm";

describe('NewRecipeForm', () => {

    const mockSignedInUser = { id: 1 };
    const mockErrors = [];

    it('renders without crashing', () => {
        render(<NewRecipeForm addRecipe={vi.fn()} signedInUser={mockSignedInUser} errors={mockErrors} />);
    });

    it('takes a snapshot', () => {
        const { asFragment } = render(<NewRecipeForm addRecipe={vi.fn()} signedInUser={mockSignedInUser} errors={mockErrors} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('correctly updates title input', () => {
        render(<NewRecipeForm addRecipe={vi.fn()} signedInUser={mockSignedInUser} errors={mockErrors} />);
        const titleInput = screen.getByLabelText(/title/i);
        fireEvent.change(titleInput, { target: { value: 'Recipe 1' } });
        expect(titleInput).toHaveValue("Recipe 1");
    });

    it('correctly updates description input', () => {
        render(<NewRecipeForm addRecipe={vi.fn()} signedInUser={mockSignedInUser} errors={mockErrors} />);
        const descriptionInput = screen.getByLabelText(/description/i);
        fireEvent.change(descriptionInput, { target: { value: 'Tasty!' } });
        expect(descriptionInput).toHaveValue("Tasty!");
    });

    it('clears formData upon formSubmit', () => {
        render(<NewRecipeForm addRecipe={vi.fn()} signedInUser={mockSignedInUser} errors={mockErrors} />);

        const titleInput = screen.getByLabelText(/title/i);
        const descriptionInput = screen.getByLabelText(/description/i);

        fireEvent.change(titleInput, { target: { value: 'Recipe 1' } });
        fireEvent.change(descriptionInput, { target: { value: 'Tasty!' } });

        expect(titleInput).toHaveValue('Recipe 1');
        expect(descriptionInput).toHaveValue('Tasty!');

        fireEvent.click(screen.getByRole('button'));

        expect(titleInput).toHaveValue('');
        expect(descriptionInput).toHaveValue('');
    });

    it('calls addRecipe when submitted', () => {
        const addRecipeMock = vi.fn();

        render(<NewRecipeForm addRecipe={addRecipeMock} signedInUser={mockSignedInUser} errors={mockErrors} />);

        const titleInput = screen.getByLabelText(/title/i);
        const descriptionInput = screen.getByLabelText(/description/i);

        fireEvent.change(titleInput, { target: { value: 'Recipe 1' } });
        fireEvent.change(descriptionInput, { target: { value: 'Tasty!' } });

        fireEvent.click(screen.getByRole('button'));

        expect(addRecipeMock).toHaveBeenCalled();
    });

    it('displays error messages when inputs are empty', () => {
        const addRecipeMock = vi.fn();
        const errorMessages = ['Title cannot be blank.', 'Description cannot be blank.'];

        render(<NewRecipeForm addRecipe={addRecipeMock} signedInUser={mockSignedInUser} errors={errorMessages} />);

        fireEvent.click(screen.getByRole('button'));

        expect(screen.getByText('Title cannot be blank.')).toBeInTheDocument();
        expect(screen.getByText('Description cannot be blank.')).toBeInTheDocument();
    });
});
