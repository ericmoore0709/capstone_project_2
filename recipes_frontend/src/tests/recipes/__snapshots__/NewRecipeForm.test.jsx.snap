// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewRecipeForm > takes a snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="text-center mt-2"
    >
      Create a Recipe
    </h2>
    <form
      class="w-50 mx-auto p-2 border"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="title"
        >
          Title
        </label>
        <input
          class="form-control"
          id="title"
          name="title"
          type="text"
          value=""
        />
        <small
          class="form-text text-muted"
        >
          Give your recipe a name!
        </small>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="description"
        >
          Description
        </label>
        <input
          class="form-control"
          id="description"
          name="description"
          type="text"
          value=""
        />
        <small
          class="form-text text-muted"
        >
          Give a light description of your delicious masterpiece!
        </small>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="image"
        >
          Image URL
        </label>
        <input
          class="form-control"
          id="image"
          name="image"
          type="text"
          value=""
        />
        <small
          class="form-text text-muted"
        >
          Show us an image of your finished meal.
        </small>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="visibility_id"
        >
          Visibility
        </label>
        <select
          class="form-select"
          id="visibility_id"
          name="visibility_id"
        >
          <option
            value="1"
          >
            Public
          </option>
          <option
            value="2"
          >
            Community
          </option>
          <option
            value="3"
          >
            Private
          </option>
        </select>
        <small
          class="form-text text-muted"
        >
          Select who can view your recipe.
        </small>
      </div>
      <div
        class="text-center mb-3"
      >
        <button
          class="w-25 btn btn-primary"
        >
          Add
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
